### ISSUE TITLE
load.bzl

### ISSUE THREAD

**Issue by avibryant** [johnynek/bazel-deps#13]
The README refers to `load.bzl` but doesn't mention that you have to write it yourself, or link to any default implementation.

If you copy the one from this repo's 3rdparty, you will get errors about the `bind` attribute being missing.

**Comment by johnynek** [comment]
added a note in the README: https://github.com/johnynek/bazel-deps/commit/2d9caaf790839ea97e14157743d4128043f77b76

**Comment by davidsantiago** [comment]
It looks like this is done. 

But I'd add that I don't understand why there's no default implementation, nor why it is written so you have to code your own version. I suspect maybe this is for flexibility for the user somehow, but I don't see it. It could also be some limitation in Bazel, or something else I'm not thinking of. In any case, it might be helpful to add the reasoning for these to the readme.

**Comment by johnynek** [comment]
It was mainly added because we wanted to control specifics like: insisting
on Sha (which the tool could do) using a possibly non-standard way of
loading jars (not just calling maven_jar).

That said, I'm not sure we are really using these features now (we should
always get a Sha and we are just using a private maven server).

I think a default would not be hard. I'll think about a clean way to do it.

**Comment by davidsantiago** [comment]
Makes sense. 

Unrelated: do you think it would be possible to get bazel-deps working as an external repo in another project? That way you could do something like bind the gen_maven_deps.sh script to a target that you can run from your own project, and perhaps write a macro to not have to remember the arguments. I've tried loading it via git_repository() this way, but it complains about the Scala rules not being found. I'm not sure if it's possible.

**Comment by johnynek** [comment]
In bazel you need to add all transitive repos to your workspace. So you
need t ok add the scala rules also as a external git repo.

After that, it should work. The only issue you might have is that I don't
know how bazel will deal with one rule modifying the BUILD files for the
build.

You certainly won't be able take this automatic (bazel automatically runs
the tool when the yaml file changes).

**Comment by davidsantiago** [comment]
Yeah, I actually forked the project and gave it a shot last night. After adding bazel-deps and the rules_scala repo as git_repository() entries in the WORKSPACE, it fails to see the scala rules: 

```
$ bazel run @com_github_johnynek_bazel_deps//src/scala/com/github/johnynek/bazel_deps:parseproject
..
WARNING: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/WORKSPACE:1: Workspace name in /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/WORKSPACE (@com_github_pgr0ss_bazel_deps) does not match the name given in the repository's definition (@com_github_johnynek_bazel_deps); this will cause a build error in future versions.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:1:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:5:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:9:1: name 'scala_repl' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:12:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:27:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:35:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:42:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:52:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:67:1: name 'scala_binary' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:81:1: name 'scala_library' is not defined.
ERROR: /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD:93:1: name 'scala_library' is not defined.
ERROR: no such target '@com_github_johnynek_bazel_deps//src/scala/com/github/johnynek/bazel_deps:parseproject': target 'parseproject' not declared in package 'src/scala/com/github/johnynek/bazel_deps' defined by /private/var/tmp/_bazel_David/aca004537b75fd97c67330379e89470b/external/com_github_johnynek_bazel_deps/src/scala/com/github/johnynek/bazel_deps/BUILD.
INFO: Elapsed time: 0.715s
ERROR: Build failed. Not running target.
```

Not sure what to make of that, everything I read in the bazel docs makes me believe that this should work, at least as far as being able to make use of the scala_* rules. There may just be something I'm not understanding here. I'll keep working on it and see if I can get anywhere. 

Maybe that's the wrong direction, though. It might be more fruitful to just package this up as a simple installable command that you can use as a utility to update your WORKSPACE/3rdparty tree. 

Maybe that's the wrong direction, though. It might be more fruitful to just package this up as a simple installable command that you can use as a utility to update your WORKSPACE/3rdparty tree.

**Comment by johnynek** [comment]
huh. that is a bazel bug. bazel-deps puts those in the prelude:
https://github.com/johnynek/bazel-deps/blob/master/tools/build_rules/prelude_bazel

it seems the prelude is ignored in remove repos.

**Comment by johnynek** [comment]
We use it as a command you run, not as something in tree.

We build the deploy jar, and just run it.

**Comment by davidsantiago** [comment]
Weird, just noticed the last paragraph in my previous post was inserted twice. Not sure how that happened. 

Anyways, very interesting about the prelude. I hadn't heard of the Bazel prelude before, and I find no mention of it on bazel.io, and only two passing references on the Google Groups. So I'm not sure what the semantics of it are supposed to be. However, once I added that prelude file to my own repo, it started seeing the rules. I agree it does feel like a bug that a remote repo does not have its prelude respected. Going to file a feature request for a little documentation on the prelude. 

I like the idea of cloning a project and the build system being able to have or get everything it needs to build and run the project (and I think bazel seems to like that sort of thing too). But if that's not possible, setting up a symlink to the deploy jar would also be just fine I think. This appears to be how buildifier works, which also messes with BUILD files.
